### DISCUSSION TITLE
Template function for relative path?

### DISCUSSION THREAD

**Discussion by 2-4601** [twpayne/chezmoi#3765] [Q&A]
Is there a template function that gives a relative path when given a base path and a target path as arguments?

I have something like this in mind:
```
$ chezmoi execute-template '{{ relativePath "/home/2-4601" "/home/2-4601/.local/share" }}'

.local/share
```

Golang's `filepath` seems to have a function like this: https://pkg.go.dev/path/filepath#Rel

**Comment by Lockszmith-GH** [comment]
how about just a replace string?

Sounds like an [XYproblem](https://xyproblem.info), please explain what you are trying to achieve.

**Reply by 2-4601** [reply]
Yeah, it can be seen as an XY problem. I am already using `replace` and that works. I was just wondering if there is a template function that mimics what filepath.Rel does because that would be more readable.

Background is that I wanted to stop using literal paths in my centralized externals file located in chezmoi root.

# Example
## Before:
```
...
[".local/share/fonts/Noto_Emoji_monochrome_variable-width.ttf"]
    type = "file"
    url = "https://github.com/google/fonts/raw/main/ofl/notoemoji/NotoEmoji%5Bwght%5D.ttf"
    refreshPeriod = "{{ mul 24 30 2 }}h"
...
```

## After:
```
{{- $data_dir := or (env "XDG_DATA_HOME") ".local/share" | replace .chezmoi.homeDir "" | trimPrefix "/" -}}
...
["{{ joinPath $data_dir "fonts/Noto_Emoji_monochrome_variable-width.ttf" }}"]
    type = "file"
    url = "https://github.com/google/fonts/raw/main/ofl/notoemoji/NotoEmoji%5Bwght%5D.ttf"
    refreshPeriod = "{{ mul 24 30 2 }}h"
...
```

-----

The externals file expects relative paths, i.e. absolute paths do not work. Another option would be to create multiple externals files straight in the appropriate directories, but I want to keep my externals in a single centralized file.

**Reply by 2-4601** [reply]
Now that I wrote the above example, I realized that I don't need the `replace`. I can just `trimPrefix .chezmoi.homeDir`, which is much more readable. But still not as nice as a `relativePath` would be :)

**Reply by Lockszmith-GH** [reply]
That's the nature of XYproblems, when you spell things out, aspects of the question either fall away or morph into the correct question 🧐 

Anyway, per [chezmoi docs](https://www.chezmoi.io/user-guide/templating/), the templating is done by [sprig](http://masterminds.github.io/sprig/), doesn't look like any of the [Path functions](http://masterminds.github.io/sprig/paths.html) have what you are asking for.

An option to simplify your files a bit might be to do all of the 'less nice' code once, during initialization and store them in `.chezmoi.toml.tmpl`

For example:
```tmpl
{{- $data_dir := osClean ( or (env "XDG_DATA_HOME") (env "USERPROFILE") ".local/share" | join "/" ) | trimPrefix ( osClean (list .chezmoi.homeDir ".:." | join "/" ) | trimSuffix ".:." ) }}

[data]
{{- if ($data_dir) }}
    data_dir = {{ $data_dir | quote }}
{{- end -}}
...
```

Then in your templates you can just use `{{ .chezmoi.data.data_dir }}`. It makes for a cleaner overall solution that spans across all your files.
After all, once the system has initialized there is no need to recalibrate.

Since you are running only once, you might be able to rely on external tools to resolve whatever you need, even if it a bit clunky or crude.

**Reply by 2-4601** [reply]
Thanks. Saving the user data path to the config file as data is a nice idea. I think it would be best to save the absolute path in that case, because it is more usable everywhere else. The externals file is an exception because it wants relative paths. So there needs to be a special case that's best to handle in the externals file itself.

**Reply by Lockszmith-GH** [reply]
Makes total sense, and also - why not both 🤭?

**Reply by twpayne** [reply]
Note that you need to be careful on Windows, as Windows uses both `\` and `/` as path separators (and a whole bunch of other path-related insanity with drive letter, shared drives, UNC paths, `C:\PROGRA~1` vs. `C:\Program Files`, etc.) so using `trimPrefix` will need testing for your particular configuration if you're using Windows.

**Reply by Lockszmith-GH** [reply]
@twpayne the / vs \ is why my version wraps the path with osClean, this way I always get the correct local-os flavor.

And I totally agree about the quirks of short vs long path - always test.
